runtimes/spiritnet/src/tests.rs

```diff
@@ -27,3 +29,87 @@ fn call_size() {
 		If the limit is too strong, maybe consider increase the limit to 300.",
 	);
 }
+
+#[test]
+fn test_derive_did_verification_relation_ctype() {
+	let c1 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3],
+	});
+	let c2 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3, 3],
+	});
+	let c3 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3, 3],
+	});
+	let c4 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 100],
+	});
+
+	let cb = Call::Utility(pallet_utility::Call::batch {
+		calls: vec![c1, c2, c3, c4],
+	});
+	assert_eq!(
+		cb.derive_verification_key_relationship(),
+		Some(did::DidVerificationKeyRelationship::AssertionMethod)
+	);
+}
+
+#[test]
+fn test_derive_did_verification_relation_fail() {
+	let c1 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3],
+	});
+	let c2 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3, 3],
+	});
+	let c3 = Call::System(frame_system::Call::remark {
+		remark: vec![0, 1, 2, 3, 3],
+	});
+	let c4 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 100],
+	});
+
+	let cb = Call::Utility(pallet_utility::Call::batch {
+		calls: vec![c1, c2, c3, c4],
+	});
+	assert_eq!(cb.derive_verification_key_relationship(), None);
+}
+
+#[test]
+fn test_derive_did_verification_relation_nested_fail() {
+	let c1 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3],
+	});
+	let c2 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 3, 3],
+	});
+	let f3 = Call::System(frame_system::Call::remark {
+		remark: vec![0, 1, 2, 3, 3],
+	});
+	let c4 = Call::Ctype(ctype::Call::add {
+		ctype: vec![0, 1, 2, 100],
+	});
+
+	let cb = Call::Utility(pallet_utility::Call::batch {
+		calls: vec![c1.clone(), c2.clone(), c4.clone()],
+	});
+
+	let cb = Call::Utility(pallet_utility::Call::batch {
+		calls: vec![c1, c2, cb, f3, c4],
+	});
+	assert_eq!(cb.derive_verification_key_relationship(), None);
```

Comment:
S: This returns `Some(AssertionMethod)` if `f3` is removed from the batch, so perhaps I would also add a (positive) test for nested calls where all calls require the same key type (also for peregrine runtime)?

Reply:
e88b802453862ef1218b44a5333cadf33c972440